internal/handler/ingest_handler.go

```diff
@@ -0,0 +1,112 @@
+package handler
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	indexer "github.com/filecoin-project/go-indexer-core"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/models"
+	"github.com/filecoin-project/storetheindex/internal/providers"
+	"github.com/filecoin-project/storetheindex/internal/syserr"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// IngestHandler provides request handling functionality for the ingest server
+// that is common to all protocols
+type IngestHandler struct {
+	indexer  indexer.Interface
+	registry *providers.Registry
+}
+
+func NewIngestHandler(indexer indexer.Interface, registry *providers.Registry) *IngestHandler {
+	return &IngestHandler{
+		indexer:  indexer,
+		registry: registry,
+	}
+}
+
+func (h *IngestHandler) DiscoverProvider(data []byte) error {
+	discoReq, err := models.ReadDiscoverRequest(data)
+	if err != nil {
+		return fmt.Errorf("connot read discover request: %s", err)
+	}
+
+	if err = h.registry.CheckSequence(discoReq.ProviderID, discoReq.Seq); err != nil {
+		return err
+	}
+
+	return h.registry.Discover(discoReq.ProviderID, discoReq.DiscoveryAddr, false)
+}
+
+func (h *IngestHandler) RegisterProvider(data []byte) error {
+	peerRec, err := models.ReadRegisterRequest(data)
+	if err != nil {
+		return fmt.Errorf("cannot read register request: %s", err)
+	}
+
+	if len(peerRec.PeerID) == 0 {
+		return errors.New("missing peer id")
+	}
+
+	if err = h.registry.CheckSequence(peerRec.PeerID, peerRec.Seq); err != nil {
+		return err
+	}
+
+	info := &providers.ProviderInfo{
+		AddrInfo: peer.AddrInfo{
+			ID:    peerRec.PeerID,
+			Addrs: peerRec.Addrs,
+		},
+	}
+	return h.registry.Register(info)
+}
+
+func (h *IngestHandler) ListProviders() ([]byte, error) {
+	infos := h.registry.AllProviderInfo()
```

Comment:
We are not populating the `lastIndex` info for providers here, right? This is data is available in the datastore, so we could directly fetch it from there: https://github.com/filecoin-project/storetheindex/blob/94fbb1becd2018a2bd83b366f01eea8740740979/internal/ingest/ingest.go#L276

Reply:
Seems like we should set the value in the registry here.  If we are not going to do that, then we should remove it from the registry and only fetch it from the datastore.  What do you think is better?

Reply:
Added TODO comment

Reply:
> The lastIndex and iastIndexTime are to record when an entire advertised index has been updated, so that the indexer can know if it can ignore advertisements. I think we probably want to update something else when individual pieces of content are updated. Maybe set a last update time for the provider?

We should link the logic in the ingestion protocol (that is where we are tracking this) with the provider info. It should be straightfoward with what we have now. Let's file an issue and decide what data we want to track. Thanks!